File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/ui/EpisodesFragment.java
/*
 * Copyright 2014 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.battlelancer.seriesguide.ui;

import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.database.Cursor;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.ActivityOptionsCompat;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.ListFragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.PopupMenu;
import com.battlelancer.seriesguide.Constants;
import com.battlelancer.seriesguide.R;
import com.battlelancer.seriesguide.adapters.EpisodesAdapter;
import com.battlelancer.seriesguide.adapters.EpisodesAdapter.OnFlagEpisodeListener;
import com.battlelancer.seriesguide.enums.EpisodeFlags;
import com.battlelancer.seriesguide.provider.SeriesGuideContract.Episodes;
import com.battlelancer.seriesguide.provider.SeriesGuideContract.ListItemTypes;
import com.battlelancer.seriesguide.provider.SeriesGuideDatabase.Tables;
import com.battlelancer.seriesguide.settings.DisplaySettings;
import com.battlelancer.seriesguide.ui.dialogs.ManageListsDialogFragment;
import com.battlelancer.seriesguide.ui.dialogs.SortDialogFragment;
import com.battlelancer.seriesguide.util.EpisodeTools;
import com.battlelancer.seriesguide.util.Utils;

/**
 * Displays a list of episodes of a season.
 */
public class EpisodesFragment extends ListFragment
        implements OnClickListener, OnFlagEpisodeListener, EpisodesAdapter.PopupMenuClickListener {

    private static final String TAG = "Episodes";

    private Constants.EpisodeSorting mSorting;

    private boolean mDualPane;

    private EpisodesAdapter mAdapter;

    private int mStartingPosition;
    private long mLastCheckedItemId;

    /**
     * All values have to be integer.
     */
    public interface InitBundle {

        String SHOW_TVDBID = "show_tvdbid";

        String SEASON_TVDBID = "season_tvdbid";

        String SEASON_NUMBER = "season_number";

        String STARTING_POSITION = "starting_position";
    }

    public static EpisodesFragment newInstance(int showId, int seasonId, int seasonNumber,
            int startingPosition) {
        EpisodesFragment f = new EpisodesFragment();

        Bundle args = new Bundle();
        args.putInt(InitBundle.SHOW_TVDBID, showId);
        args.putInt(InitBundle.SEASON_TVDBID, seasonId);
        args.putInt(InitBundle.SEASON_NUMBER, seasonNumber);
        args.putInt(InitBundle.STARTING_POSITION, startingPosition);
        f.setArguments(args);

        return f;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
            Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_episodes, container, false);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        loadSortOrder();

        // listen to changes to the sorting preference
        final SharedPreferences prefs = PreferenceManager
                .getDefaultSharedPreferences(getActivity());
        prefs.registerOnSharedPreferenceChangeListener(mSortOrderChangeListener);

        // Check to see if we have a frame in which to embed the details
        // fragment directly in the containing UI.
        View pager = getActivity().findViewById(R.id.pagerEpisodes);
        mDualPane = pager != null && pager.getVisibility() == View.VISIBLE;

        if (mDualPane) {
            getListView().setChoiceMode(ListView.CHOICE_MODE_SINGLE);
            mStartingPosition = getArguments().getInt(InitBundle.STARTING_POSITION);
        } else {
            mStartingPosition = -1;
        }
        mLastCheckedItemId = -1;

        mAdapter = new EpisodesAdapter(getActivity(), null, 0, this, this);
        setListAdapter(mAdapter);

        getLoaderManager().initLoader(EpisodesActivity.EPISODES_LOADER_ID, null, mLoaderCallbacks);

        setHasOptionsMenu(true);
    }

    private int getShowId() {
        return getArguments().getInt(InitBundle.SHOW_TVDBID);
    }

    private int getSeasonId() {
        return getArguments().getInt(InitBundle.SEASON_TVDBID);
    }

    private int getSeasonNumber() {
        return getArguments().getInt(InitBundle.SEASON_NUMBER);
    }

    /**
     * Display the episode at the given position in a detail pane or if not available in a new
     * activity.
     */
    private void showDetails(View view, int position) {
        if (mDualPane) {
            EpisodesActivity activity = (EpisodesActivity) getActivity();
            activity.setCurrentPage(position);
            setItemChecked(position);
        } else {
            int episodeId = (int) getListView().getItemIdAtPosition(position);

            Intent intent = new Intent();
            intent.setClass(getActivity(), EpisodesActivity.class);
            intent.putExtra(EpisodesActivity.InitBundle.EPISODE_TVDBID, episodeId);

            ActivityCompat.startActivity(getActivity(), intent,
                    ActivityOptionsCompat
                            .makeScaleUpAnimation(view, 0, 0, view.getWidth(), view.getHeight())
                            .toBundle()
            );
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        loadSortOrder();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        // stop listening to sort pref changes
        final SharedPreferences prefs = PreferenceManager
                .getDefaultSharedPreferences(getActivity());
        prefs.unregisterOnSharedPreferenceChangeListener(mSortOrderChangeListener);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.episodelist_menu, menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int itemId = item.getItemId();
        if (itemId == R.id.menu_epsorting) {
            fireTrackerEvent("Sort");
            showSortDialog();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onListItemClick(ListView l, View view, int position, long id) {
        showDetails(view, position);
    }

    @Override
    public void onPopupMenuClick(View v, final int episodeTvdbId, final int episodeNumber,
            final long releaseTimeMs, final int watchedFlag, final boolean isCollected) {
        PopupMenu popupMenu = new PopupMenu(v.getContext(), v);
        popupMenu.inflate(R.menu.episodes_popup_menu);

        Menu menu = popupMenu.getMenu();
        menu.findItem(R.id.menu_action_episodes_collection_add).setVisible(!isCollected);
        menu.findItem(R.id.menu_action_episodes_collection_remove).setVisible(isCollected);
        boolean isWatched = EpisodeTools.isWatched(watchedFlag);
        menu.findItem(R.id.menu_action_episodes_watched).setVisible(!isWatched);
        menu.findItem(R.id.menu_action_episodes_not_watched).setVisible(isWatched);
        boolean isSkipped = EpisodeTools.isSkipped(watchedFlag);
        menu.findItem(R.id.menu_action_episodes_skip).setVisible(!isWatched && !isSkipped);
        menu.findItem(R.id.menu_action_episodes_dont_skip).setVisible(isSkipped);

        popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                switch (item.getItemId()) {
                    case R.id.menu_action_episodes_watched: {
                        onFlagEpisodeWatched(episodeTvdbId, episodeNumber, true);
                        fireTrackerEventContextMenu("Flag watched");
                        return true;
                    }
                    case R.id.menu_action_episodes_not_watched: {
                        onFlagEpisodeWatched(episodeTvdbId, episodeNumber, false);
                        fireTrackerEventContextMenu("Flag unwatched");
                        return true;
                    }
                    case R.id.menu_action_episodes_collection_add: {
                        onFlagEpisodeCollected(episodeTvdbId, episodeNumber, true);
                        fireTrackerEventContextMenu("Flag collected");
                        return true;
                    }
                    case R.id.menu_action_episodes_collection_remove: {
                        onFlagEpisodeCollected(episodeTvdbId, episodeNumber, false);
                        fireTrackerEventContextMenu("Flag uncollected");
                        return true;
                    }
                    case R.id.menu_action_episodes_skip: {
                        onFlagEpisodeSkipped(episodeTvdbId, episodeNumber, true);
                        fireTrackerEventContextMenu("Flag skipped");
                        return true;
                    }
                    case R.id.menu_action_episodes_dont_skip: {
                        onFlagEpisodeSkipped(episodeTvdbId, episodeNumber, false);
                        fireTrackerEvent("Flag not skipped");
                        return true;
                    }
                    case R.id.menu_action_episodes_watched_previous: {
                        onMarkUntilHere(releaseTimeMs);
                        fireTrackerEventContextMenu("Flag previously aired");
                        return true;
                    }
                    case R.id.menu_action_episodes_manage_lists: {
                        ManageListsDialogFragment.showListsDialog(episodeTvdbId,
                                ListItemTypes.EPISODE,
                                getFragmentManager());
                        fireTrackerEventContextMenu("Manage lists");
                        return true;
                    }
                }
                return false;
            }
        });
        popupMenu.show();
    }

    @Override
    public void onFlagEpisodeWatched(int episodeTvdbId, int episode, boolean isWatched) {
        EpisodeTools.episodeWatched(getActivity(), getShowId(), episodeTvdbId, getSeasonNumber(),
                episode, isWatched ? EpisodeFlags.WATCHED : EpisodeFlags.UNWATCHED);
    }

    public void onFlagEpisodeSkipped(int episodeTvdbId, int episode, boolean isSkipped) {
        EpisodeTools.episodeWatched(getActivity(), getShowId(), episodeTvdbId, getSeasonNumber(),
                episode, isSkipped ? EpisodeFlags.SKIPPED : EpisodeFlags.UNWATCHED);
    }

    public void onFlagEpisodeCollected(int episodeTvdbId, int episode, boolean isCollected) {
        EpisodeTools.episodeCollected(getActivity(), getShowId(), episodeTvdbId, getSeasonNumber(),
                episode, isCollected);
    }

    private void onMarkUntilHere(long episodeFirstReleaseMs) {
        EpisodeTools.episodeWatchedPrevious(getActivity(), getShowId(), episodeFirstReleaseMs);
    }

    private LoaderManager.LoaderCallbacks<Cursor> mLoaderCallbacks
            = new LoaderManager.LoaderCallbacks<Cursor>() {
        @Override
        public Loader<Cursor> onCreateLoader(int id, Bundle args) {
            return new CursorLoader(getActivity(),
                    Episodes.buildEpisodesOfSeasonWithShowUri(String.valueOf(getSeasonId())),
                    EpisodesQuery.PROJECTION, null, null, mSorting.query());
        }

        @Override
        public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
            mAdapter.swapCursor(data);
            // set an initial checked item
            if (mStartingPosition != -1) {
                setItemChecked(mStartingPosition);
                mStartingPosition = -1;
            }
            // correctly restore the last checked item
            else if (mLastCheckedItemId != -1) {
                setItemChecked(mAdapter.getItemPosition(mLastCheckedItemId));
                mLastCheckedItemId = -1;
            }
        }

        @Override
        public void onLoaderReset(Loader<Cursor> loader) {
            mAdapter.swapCursor(null);
        }
    };

    public interface EpisodesQuery {

        String[] PROJECTION = new String[] {
                Tables.EPISODES + "." + Episodes._ID, Episodes.WATCHED, Episodes.TITLE,
                Episodes.NUMBER, Episodes.FIRSTAIREDMS, Episodes.DVDNUMBER,
                Episodes.ABSOLUTE_NUMBER, Episodes.COLLECTED
        };

        int _ID = 0;

        int WATCHED = 1;

        int TITLE = 2;

        int NUMBER = 3;

        int FIRSTAIREDMS = 4;

        int DVDNUMBER = 5;

        int ABSOLUTE_NUMBER = 6;

        int COLLECTED = 7;
    }

    private void loadSortOrder() {
        mSorting = DisplaySettings.getEpisodeSortOrder(getActivity());
    }

    private void showSortDialog() {
        FragmentManager fm = getFragmentManager();
        SortDialogFragment sortDialog = SortDialogFragment.newInstance(R.array.epsorting,
                R.array.epsortingData, mSorting.index(),
                DisplaySettings.KEY_EPISODE_SORT_ORDER, R.string.pref_episodesorting);
        sortDialog.show(fm, "fragment_sort");
    }

    private final OnSharedPreferenceChangeListener mSortOrderChangeListener
            = new OnSharedPreferenceChangeListener() {
        @Override
        public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
            if (DisplaySettings.KEY_EPISODE_SORT_ORDER.equals(key)) {
                onSortOrderChanged();
            }
        }
    };

    private void onSortOrderChanged() {
        loadSortOrder();

        mLastCheckedItemId = getListView().getItemIdAtPosition(
                getListView().getCheckedItemPosition());
        getLoaderManager().restartLoader(EpisodesActivity.EPISODES_LOADER_ID, null,
                mLoaderCallbacks);

        Utils.trackCustomEvent(getActivity(), TAG, "Sorting", mSorting.name());
    }

    /**
     * Highlight the given episode in the list.
     */
    public void setItemChecked(int position) {
        ListView list = getListView();
        list.setItemChecked(position, true);
        if (position <= list.getFirstVisiblePosition()
                || position >= list.getLastVisiblePosition()) {
            list.smoothScrollToPosition(position);
        }
    }

    @Override
    public void onClick(View v) {
        getActivity().openContextMenu(v);
    }

    private void fireTrackerEvent(String label) {
        Utils.trackAction(getActivity(), TAG, label);
    }

    private void fireTrackerEventContextMenu(String label) {
        Utils.trackContextMenu(getActivity(), TAG, label);
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/shows/episodes/EpisodesFragment.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2018-2024 Uwe Trottmann

package com.battlelancer.seriesguide.shows.episodes

import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.widget.PopupMenu
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.LinearLayoutManager
import com.battlelancer.seriesguide.R
import com.battlelancer.seriesguide.databinding.FragmentEpisodesBinding
import com.battlelancer.seriesguide.ui.dialogs.SingleChoiceDialogFragment
import com.battlelancer.seriesguide.ui.widgets.SgFastScroller
import com.battlelancer.seriesguide.util.ThemeUtils
import com.battlelancer.seriesguide.util.Utils
import com.battlelancer.seriesguide.util.safeShow

/**
 * Displays a list of episodes of a season.
 */
class EpisodesFragment : Fragment() {

    private var seasonId: Long = 0
    private var startingPosition: Int = 0
    private var scrollToCheckedItemOnDataRefresh = false
    private var watchedAllEpisodes: Boolean = false
    private var collectedAllEpisodes: Boolean = false

    private var binding: FragmentEpisodesBinding? = null
    private lateinit var adapter: EpisodesAdapter
    private val model by viewModels<EpisodesViewModel> {
        EpisodesViewModelFactory(requireActivity().application, seasonId)
    }

    companion object {

        private const val ARG_LONG_SEASON_ID = "season_id"
        private const val ARG_INT_STARTING_POSITION = "starting_position"

        private const val CONTEXT_WATCHED_ALL = 1
        private const val CONTEXT_WATCHED_NONE = 2
        private const val CONTEXT_COLLECTED_ALL = 3
        private const val CONTEXT_COLLECTED_NONE = 4

        @JvmStatic
        fun newInstance(
            seasonId: Long,
            startingPosition: Int
        ): EpisodesFragment {
            return EpisodesFragment().apply {
                arguments = Bundle().apply {
                    putLong(ARG_LONG_SEASON_ID, seasonId)
                    putInt(ARG_INT_STARTING_POSITION, startingPosition)
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.let {
            seasonId = it.getLong(ARG_LONG_SEASON_ID)
            startingPosition = it.getInt(ARG_INT_STARTING_POSITION)
        } ?: throw IllegalArgumentException("Missing arguments")
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val binding = FragmentEpisodesBinding.inflate(inflater, container, false)
        this.binding = binding
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding?.also { binding ->
            ThemeUtils.applyBottomPaddingForNavigationBar(binding.recyclerViewEpisodes)

            binding.imageViewEpisodesWatched.setImageResource(R.drawable.ic_watch_all_black_24dp)
            binding.imageViewEpisodesCollected.setImageResource(R.drawable.ic_collect_all_black_24dp)

            adapter = EpisodesAdapter(requireActivity(), episodesListItemClickListener)
            adapter.selectedItemId = model.selectedItemId

            binding.recyclerViewEpisodes.also {
                it.layoutManager = LinearLayoutManager(requireContext())
                it.adapter = adapter
                SgFastScroller(requireContext(), it)
            }
        }

        // listen to changes to the sorting preference
        PreferenceManager.getDefaultSharedPreferences(requireActivity()).apply {
            registerOnSharedPreferenceChangeListener(onSortOrderChangedListener)
        }

        model.episodeCounts.observe(viewLifecycleOwner) { result ->
            setWatchedToggleState(result.unwatchedEpisodes)
            setCollectedToggleState(result.uncollectedEpisodes)
        }
        model.episodes.observe(viewLifecycleOwner) { episodes ->
            adapter.submitList(episodes) {
                // set and scroll to an initial checked item
                if (savedInstanceState == null && startingPosition != -1) {
                    setItemChecked(startingPosition)
                    startingPosition = -1
                } else if (scrollToCheckedItemOnDataRefresh) {
                    val position = adapter.getPositionForId(adapter.selectedItemId)
                    if (position != -1) {
                        binding?.recyclerViewEpisodes?.smoothScrollToPosition(position)
                    }
                    scrollToCheckedItemOnDataRefresh = false
                }
            }
            // update count state every time data changes
            model.updateCounts()
        }

        requireActivity().addMenuProvider(
            optionsMenuProvider,
            viewLifecycleOwner,
            Lifecycle.State.RESUMED
        )
    }

    /**
     * Display the episode at the given position in the detail pane, highlight it in the list.
     */
    private fun showDetails(position: Int) {
        val activity = requireActivity() as EpisodesActivity
        activity.setCurrentPage(position)
        // Note: page change listener will update checked episode.
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    override fun onDestroy() {
        super.onDestroy()

        // stop listening to sort pref changes
        PreferenceManager.getDefaultSharedPreferences(requireActivity()).apply {
            unregisterOnSharedPreferenceChangeListener(onSortOrderChangedListener)
        }
    }

    private val optionsMenuProvider = object : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.episodelist_menu, menu)
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
            return when (menuItem.itemId) {
                R.id.menu_epsorting -> {
                    showSortDialog()
                    true
                }

                else -> false
            }
        }
    }

    private val episodesListItemClickListener = object : EpisodesAdapter.ItemClickListener {

        override fun onItemClick(position: Int) {
            showDetails(position)
        }

        /**
         * If episode is watched, asks whether to watch again or set not watched.
         * Otherwise, flags as watched.
         */
        override fun onWatchedBoxClick(anchor: View, episodeId: Long, watchedFlag: Int) {
            val watched = EpisodeTools.isWatched(watchedFlag)
            if (watched) {
                PopupMenu(anchor.context, anchor).apply {
                    inflate(R.menu.watched_popup_menu)
                    setOnMenuItemClickListener { item: MenuItem ->
                        val itemId = item.itemId
                        if (itemId == R.id.watched_popup_menu_watch_again) {
                            // Multiple plays are for supporters only.
                            if (!Utils.hasAccessToX(requireContext())) {
                                Utils.advertiseSubscription(requireContext())
                            } else {
                                onFlagEpisodeWatched(episodeId, true)
                            }
                        } else if (itemId == R.id.watched_popup_menu_set_not_watched) {
                            onFlagEpisodeWatched(episodeId, false)
                        }
                        true
                    }
                }.show()
            } else {
                onFlagEpisodeWatched(episodeId, true)
            }
        }


        override fun onMoreOptionsClick(
            anchor: View,
            episodeId: Long,
            episodeNumber: Int,
            releaseTimeMs: Long,
            watchedFlag: Int,
            isCollected: Boolean
        ) {
            PopupMenu(anchor.context, anchor).apply {
                inflate(R.menu.episodes_popup_menu)
                menu.apply {
                    findItem(R.id.menu_action_episodes_collection_add).isVisible = !isCollected
                    findItem(R.id.menu_action_episodes_collection_remove).isVisible =
                        isCollected
                    val isWatched = EpisodeTools.isWatched(watchedFlag)
                    findItem(R.id.menu_action_episodes_watched).isVisible = !isWatched
                    findItem(R.id.menu_action_episodes_not_watched).isVisible = isWatched
                    findItem(R.id.menu_action_episodes_watched_up_to).isVisible = !isWatched
                    val isSkipped = EpisodeTools.isSkipped(watchedFlag)
                    findItem(R.id.menu_action_episodes_skip).isVisible =
                        !isWatched && !isSkipped
                    findItem(R.id.menu_action_episodes_dont_skip).isVisible = isSkipped
                }
                setOnMenuItemClickListener { item ->
                    when (item.itemId) {
                        R.id.menu_action_episodes_watched -> {
                            onFlagEpisodeWatched(episodeId, true)
                            true
                        }

                        R.id.menu_action_episodes_not_watched -> {
                            onFlagEpisodeWatched(episodeId, false)
                            true
                        }

                        R.id.menu_action_episodes_collection_add -> {
                            onFlagEpisodeCollected(episodeId, true)
                            true
                        }

                        R.id.menu_action_episodes_collection_remove -> {
                            onFlagEpisodeCollected(episodeId, false)
                            true
                        }

                        R.id.menu_action_episodes_skip -> {
                            onFlagEpisodeSkipped(episodeId, true)
                            true
                        }

                        R.id.menu_action_episodes_dont_skip -> {
                            onFlagEpisodeSkipped(episodeId, false)
                            true
                        }

                        R.id.menu_action_episodes_watched_up_to -> {
                            EpisodeWatchedUpToDialog.newInstance(
                                model.showId,
                                releaseTimeMs,
                                episodeNumber
                            ).safeShow(parentFragmentManager, "EpisodeWatchedUpToDialog")
                            true
                        }

                        else -> false
                    }
                }
            }.show()
        }
    }

    private fun onFlagEpisodeWatched(episodeId: Long, isWatched: Boolean) {
        EpisodeTools.episodeWatched(
            requireContext(),
            episodeId,
            if (isWatched) EpisodeFlags.WATCHED else EpisodeFlags.UNWATCHED
        )
    }

    private fun onFlagEpisodeSkipped(episodeId: Long, isSkipped: Boolean) {
        EpisodeTools.episodeWatched(
            requireContext(),
            episodeId,
            if (isSkipped) EpisodeFlags.SKIPPED else EpisodeFlags.UNWATCHED
        )
    }

    private fun onFlagEpisodeCollected(episodeId: Long, isCollected: Boolean) {
        EpisodeTools.episodeCollected(
            requireContext(),
            episodeId,
            isCollected
        )
    }

    private fun showSortDialog() {
        SingleChoiceDialogFragment.show(
            parentFragmentManager,
            R.array.epsorting,
            R.array.epsortingData,
            EpisodesSettings.getEpisodeSortOrder(requireActivity()).index(),
            EpisodesSettings.KEY_EPISODE_SORT_ORDER,
            R.string.pref_episodesorting,
            "episodeSortOrderDialog"
        )
    }

    private val onSortOrderChangedListener = OnSharedPreferenceChangeListener { _, key ->
        if (EpisodesSettings.KEY_EPISODE_SORT_ORDER == key) {
            scrollToCheckedItemOnDataRefresh = true
            model.updateOrder()
        }
    }

    /**
     * Highlight the given episode in the list.
     */
    fun setItemChecked(position: Int) {
        binding?.recyclerViewEpisodes?.also {
            model.selectedItemId = adapter.selectItem(position)
            it.smoothScrollToPosition(position)
        }
    }

    private fun setWatchedToggleState(unwatchedEpisodes: Int) {
        watchedAllEpisodes = unwatchedEpisodes == 0
        binding?.imageViewEpisodesWatched?.apply {
            // using vectors is safe because it will be an AppCompatImageView
            contentDescription = if (watchedAllEpisodes) {
                setImageResource(R.drawable.ic_watched_all_24dp)
                getString(R.string.unmark_all)
            } else {
                setImageResource(R.drawable.ic_watch_all_black_24dp)
                getString(R.string.mark_all)
            }
            // set onClick listener not before here to avoid unexpected actions
            setOnClickListener(watchedAllClickListener)
        }
    }

    private val watchedAllClickListener = View.OnClickListener { view ->
        PopupMenu(view.context, view).apply {
            if (!watchedAllEpisodes) {
                menu.add(0, CONTEXT_WATCHED_ALL, 0, R.string.mark_all)
            }
            menu.add(0, CONTEXT_WATCHED_NONE, 0, R.string.unmark_all)
            setOnMenuItemClickListener { item ->
                when (item.itemId) {
                    CONTEXT_WATCHED_ALL -> {
                        EpisodeTools.seasonWatched(
                            context,
                            seasonId,
                            EpisodeFlags.WATCHED
                        )
                        true
                    }

                    CONTEXT_WATCHED_NONE -> {
                        EpisodeTools.seasonWatched(
                            context,
                            seasonId,
                            EpisodeFlags.UNWATCHED
                        )
                        true
                    }

                    else -> false
                }
            }
        }.show()
    }

    private fun setCollectedToggleState(uncollectedEpisodes: Int) {
        collectedAllEpisodes = uncollectedEpisodes == 0
        binding?.imageViewEpisodesCollected?.apply {
            // using vectors is safe because it will be an AppCompatImageView
            contentDescription = if (collectedAllEpisodes) {
                setImageResource(R.drawable.ic_collected_all_24dp)
                getString(R.string.uncollect_all)
            } else {
                setImageResource(R.drawable.ic_collect_all_black_24dp)
                getString(R.string.collect_all)
            }
            // set onClick listener not before here to avoid unexpected actions
            setOnClickListener(collectedAllClickListener)
        }
    }

    private val collectedAllClickListener = View.OnClickListener { view ->
        PopupMenu(view.context, view).apply {
            if (!collectedAllEpisodes) {
                menu.add(0, CONTEXT_COLLECTED_ALL, 0, R.string.collect_all)
            }
            menu.add(0, CONTEXT_COLLECTED_NONE, 0, R.string.uncollect_all)
            setOnMenuItemClickListener { item ->
                when (item.itemId) {
                    CONTEXT_COLLECTED_ALL -> {
                        EpisodeTools.seasonCollected(
                            context,
                            seasonId,
                            true
                        )
                        true
                    }

                    CONTEXT_COLLECTED_NONE -> {
                        EpisodeTools.seasonCollected(
                            context,
                            seasonId,
                            false
                        )
                        true
                    }

                    else -> false
                }
            }
        }.show()
    }

}
